internal/ingress/controller/controller_test.go

```diff
@@ -20,13 +20,113 @@ import (
 	"crypto/x509"
 	"crypto/x509/pkix"
 	"encoding/asn1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/ingress-nginx/internal/ingress/annotations"
+	"k8s.io/ingress-nginx/internal/ingress/annotations/canary"
 	"testing"
 
 	extensions "k8s.io/api/extensions/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/ingress-nginx/internal/ingress"
 )
 
+func TestMergeVirtualBackends(t *testing.T) {
+	testCases := map[string]struct {
+		ingress                  *extensions.Ingress
+		upstreams                map[string]*ingress.Backend
+		servers                  map[string]*ingress.Server
+		expNumVirtualBackends    int
+		expVirtualBackendContent ingress.VirtualBackend
+		expNumLocations          int
+	}{
+		"virtual backend has no server and embeds into matching real backend": {
+			&extensions.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "example",
+				},
+				Spec: extensions.IngressSpec{
+					Rules: []extensions.IngressRule{
+						{
+							Host: "example.com",
+							IngressRuleValue: extensions.IngressRuleValue{
+								HTTP: &extensions.HTTPIngressRuleValue{
+									Paths: []extensions.HTTPIngressPath{
+										{
+											Path: "/",
+											Backend: extensions.IngressBackend{
+												ServiceName: "http-svc-canary",
+												ServicePort: intstr.IntOrString{
+													Type:   intstr.Int,
+													IntVal: 80,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			map[string]*ingress.Backend{
+				"example-http-svc-80": {
+					Name:    "example-http-svc-80",
+					Virtual: false,
+				},
+				"example-http-svc-canary-80": {
+					Name:    "example-http-svc-canary-80",
+					Virtual: true,
+				},
+			},
+			map[string]*ingress.Server{
+				"example.com": {
+					Hostname: "example.com",
+					Locations: []*ingress.Location{
+						{
+							Path:    "/",
+							Backend: "example-http-svc-80",
+						},
+					},
+				},
```

Comment:
can you create another server here where the hostname is foo.bar and assert that virtual backend won't match it

Reply:
Added that test to ensure a 1:1 match by host name